core/src/environment.rs

```diff
@@ -81,7 +83,16 @@ impl<K: Hash + Eq, V: PartialEq> Environment<K, V> {
 
     /// Tries to find the value of a key in the Environment.
     pub fn get(&self, key: &K) -> Option<&V> {
-        self.iter_layers().find_map(|hmap| hmap.get(key))
+        increment!("Environment::get");
+
+        let mut layer_count = 0;
+        let r = self.iter_layers().find_map(|hmap| {
+            sample!("Environment.hashmap_size_get", hmap.len() as f64);
+            layer_count += 1;
+            hmap.get(key)
+        });
```

Comment:
I don't know if that's objectively better, but if you despise `mut`, here is a variant:

```suggestion
        let (layer_count, r) = self.iter_layers().enumerate().find_map(|(count, hmap)| {
            sample!("Environment.hashmap_size_get", hmap.len() as f64);
            Some((count, hmap.get(key)?))
        });
```

Reply:
I was really looking for something where I'm confident that the compiler can eliminate the dead store with minimal analysis in case `metrics` are disabled. Apparently hiding expressions behind `#[cfg(feature = "metrics")]` is experimental 😞